### ISSUE TITLE
Revert or Revise 4.x invalid-name behavior

### ISSUE THREAD

**Issue by avylove** [pylint-dev/pylint#10700]
### Current problem

The new behavior introduced in 4.0.0 via #3585, where any variable that is reassigned is considered a global rather than a constant, is non-intuitive and does not seem Pythonic.

I've provided a real world example below. The best solution to get past the error, outside of disabling the check locally, seems to be to move the logic to a function. But I doubt the logic as written would seem unexpected or inappropriate to an experienced programmer. And a function would seem like overkill for simple logic used only once.

```python
"""Simple Pylint Test"""

import platform
import sys

if platform.system() == 'Windows':
    SERIES_STD = ' ▌█'
else:
    SERIES_STD = ' ▏▎▍▌▋▊▉█'

try:
    SERIES_STD.encode(sys.__stdout__.encoding)
except UnicodeEncodeError:
    SERIES_STD = ' |'
except (AttributeError, TypeError):
    pass
```

The justification for the change seem to be based on the [argument](https://github.com/pylint-dev/pylint/issues/3585#issuecomment-1433774348) that PEP8 acknowledges globals exist and should use the same style as regular variables. The problem here is globals are a code smell and very rarely used (or needed) in well-written code. Even the part of PEP8 referenced in that comment is mainly about how to keep globals contained within the current module.

I did not see any examples in #3585 that showed code that belonged at the module level to use as a counterpoint.

### Desired solution

I think a more reasonable argument, would be that variables assigned at the module level are considered constants unless guarded by ``if __name__ == '__main__':``, where use of regular variables is reasonable. While this will not work in all cases, it is more likely to produce well-written code.

A more comprehensive check for globals would be to see if they have been reassigned within a function/method via the globals keyword. In this case, the author has been explicit that these are globals. Perhaps this list can be captured and then assignment nodes checked against it within the constants logic?

**Comment by DanielNoord** [comment]
This is just a duplicate of #10697, where apparently you didn't agree with the reason for closing the issue.

I believe @jacobtylerwalls gave some guide advice on how to solve this particular warning for the given example.
`pylint` is used for thousands of codebases and its behaviour needs to be "valid" for most. I personally still believe that the change is a better default behaviour and that for examples like the one you gave you'd either need to slightly refactor or ignore the warning.

**Comment by avylove** [comment]
@DanielNoord  #10697 was submitted as a bug report (closed without referencing the change), this is a feature request. 

Correct, Pylint needs to be valid for most. I don't think this change is valid for most.  Pylint should also result in improved code quality, not encourage poor programming practices.

#3585 reversed a decade(?) of precedent, without a Boolean option to reverse it, on an incomplete argument without any real world examples. This leaves users with some options, none ideal:

- Ignore locally
- Disable the check
- Unnecessarily refactor the code
- Implement a custom checker that uses the old logic
- Switch to another linter

So, I would argue, **at a minimum, this change should have a configuration option to restore the old behavior.**

However, I think the original change was unnecessary and encourages poor coding practices.

To start, all but one example in #3585 use some variation of

```python
for i in (0, 1, 2):
    variable = i
```

This is a contrived example, that should be flagged by Pylint, but is not, for leaving `i` as a dangling global. Though Pylint will raise ``redefined-outer-name`` if `i` is also used within a function or method. Perhaps someone can provide a real world example that would justify this change?

Going back to what appears to be the [argument](https://github.com/pylint-dev/pylint/issues/3585#issuecomment-1433774348) that seemed to hold the most weight in this discussion.

The argument seems to be that PEP8 acknowledges module-level variables that are not constants. This is under the heading [Global Variable Names](https://peps.python.org/pep-0008/#global-variable-names). So let's look at that.

1. Per PEP 8, these should be used only within the module they are define and should be excluded from ``__all__`` or start with an underscore to prevent leaking them to other modules. I don't believe Pylint is checking for this.

2. A well-written module, or even a single file script, should have all code not intended to be run at initialization guarded, either by placing it in functions or classes, or with a dunder name conditional.

3. Variables that don't change after module initialization (which may require non-exclusive assignment, like the one in the initial comment), are constants, and should be in all caps.

4. Variables that do change, should not do so in an unguarded manor. That is, if you are defining a global, changing that global should require calling a function that uses the ``global`` keyword to change the value. Otherwise, the value would just change during module initialization. Therefore, legitimate globals (or "module-level variables") defined outside of a dunder name guard would always be identifiable via use of the ``global`` keyword within a function or method. Note, Pylint recognizes use of globals as generally bad practice and, by default, raises ``global-statement`` when the global keyword is used.

**TL/DR**

The new behavior encourages poor programming practices. If it was an appropriate change, someone should be able to provide a real world example.

If the new behavior is kept, a configuration option should be added to restore the classic behavior.

**Comment by jacobtylerwalls** [comment]
Thanks for opening a new issue. I'm sensitive to the possibility that the footprint of the changes in #3585 was too wide.

We had three options for that ask ("I have module-level names that I wish to spell in snake_case because they are variables"):
1. wontfix ("don't do that")
2. let module-level names pass against _either_ constant or variable regexes, but not anything else (@Pierre-Sassoulas suggested this a few times)
3. count the assignments, and for n=1 (or exclusive) assignments, apply the constant regex, for n>1 assignments, apply the variable regex, and add liberal exceptions for lists, sets, and objects

We went with (3.) because (2.) would be a regression for people who did really depend on strict linting for constant-style names--suddenly you could start mixing SCREAMING_SNAKE_CASE and snake_case in your module-level names.

I now see we had an option 4:
4. keep the analysis from 3, but define a new naming style (module-level-variable) to use in those cases, and default it to the regex that had been used for a decade-plus (the const regex), forcing the behavior change to be **opt-in** by setting your module-level-variable regex to be the same as your variable regex.

4. I think would provide the level of configurability/opt-inness you're looking for without adding new booleans (e.g. "--classic").

Would you be open to that? I'd be happy to help pushing a 4.1 out the door with that change.

**Comment by avylove** [comment]
@jacobtylerwalls I think a separate naming style would be helpful, and increases flexibility. If it it makes the new 4.0 behavior opt-in, then I support it.

This would let someone set the module-level-variable regex to be the same as their variable regex if they chose, or, to more strictly follow PEP8, set it to require the names start with an underscore.

**Comment by Pierre-Sassoulas** [comment]
The given example is not a "constant", it's a module variable with two assignments and 4 possible values, with a control flow convoluted enough that I had to ask myself if SERIES_STD is always defined. If it was a function it would have a mccabe rating of 5 and require a dozen tests to be entirely covered (but it would be easier to test than a module level variable). 

While I believe more regex options might be the way to go for pylint... Adding a regex options will introduce negative side effects that we also need to consider. What happens if someone already use the current regex and we introduce the new one ? It's going to be complex, and I don't think that much work is warranted, and I think that it's a case of every bugs becoming a feature if they are released long enough (https://xkcd.com/1172/). So -0.5 from me.

**Comment by avylove** [comment]
@Pierre-Sassoulas, that's a very strict definition of a constant and an exaggeration of the complexity of the code.

My definition of a constant is the value is determined at module initialization and never changes once determined. In this case, SERIES_STD has 3 possible values depending on the environment it's running in. Once that value is determined, it never changes.

For context, the value of SERIES_STD depends on which characters are supported by the encoding used on the terminal. Windows terminals use CP437, which only has 1/2 and full block characters, while UTF-8 and some other Unicode encoding have block characters in 1/8 increments. If the terminal uses some other encoding, like ANSII, then no block characters are supported.

Based on what has been implemented, your definition seems to be that constants need to assigned exclusively. So let's look at that. There are two ways this can be done with this example. The first is by repeating the try block.


```python

"""Simple Pylint Test"""


import platform
import sys


if platform.system() == 'Windows':
    try:
        ' ▌█'.encode(sys.__stdout__.encoding)
    except UnicodeEncodeError:
        SERIES_STD = ' |'
    except (AttributeError, TypeError):
        SERIES_STD = ' ▌█'
    else:
        SERIES_STD = ' ▌█'
else:
    try:
        ' ▏▎▍▌▋▊▉█'.encode(sys.__stdout__.encoding)
    except UnicodeEncodeError:
        SERIES_STD = ' |'
    except (AttributeError, TypeError):
        SERIES_STD = ' ▏▎▍▌▋▊▉█'
    else:
        SERIES_STD = ' ▏▎▍▌▋▊▉█'
```

Pylint 4.0.2 runs cleanly, but you can see this is much harder to read and understand.

The second is to create a function.

```python
"""Simple Pylint Test"""

import platform
import sys


def _determine_default_series():

    if platform.system() == 'Windows':
        series = ' ▌█'
    else:
        series = ' ▏▎▍▌▋▊▉█'

    try:
        series.encode(sys.__stdout__.encoding)
    except UnicodeEncodeError:
        series = ' |'
    except (AttributeError, TypeError):
        pass

    return series


SERIES_STD = _determine_default_series()
```

Pylint 4.0.2 runs cleanly again. I see 4 negatives and 1 positive with this.
For the negatives:
  - It reduces readability (though not as bad as the previous example).
  - It uses more memory, since ``_determine_default_series()`` lives for the rest of the program.
  - The constant assignment is no longer at the top of the file.

The only positive is it's easier to test.

To your point of someone using the 4.0 regex. Yes, that's possible, but the previous behavior existed for much longer and was changed without identifying a valid real world example. I've brought this up several times now, and still no one has provided an example of well-written code that uses the new behavior. Additionally, according to PEP8, the current regex is incomplete and should require module level variables that aren't constants to begin with an underscore or be excluded from ``__all__``.

**Comment by Pierre-Sassoulas** [comment]
https://claude.ai/share/8e161f8c-24d5-4256-8d45-c4930745e922

2 maintainers thumbs-upped my previous -0.5 so I'm going to close as won't do. If you don't like this you can:

- Disable locally with ``# pylint: disable=invalid-name``
- Disable globally from the configuration with ``disable=invalid-name``
- [Play with the invalid-name regex](https://pylint.readthedocs.io/en/stable/user_guide/messages/convention/invalid-name.html#cmdoption-variable-naming-style) 
- Refactor to use a function to assign only once
- Implement your own checker with the old logic

**Comment by avylove** [comment]
@Pierre-Sassoulas it's your project and time, so you can do what you want, but I think the lack of a single example of where the new behavior is valid is very telling. Add to that the regex for the new behavior does not conform to the section of PEP8 that was used as a justification for the change, then it become unclear what standards Pylint is adhering to.

**Comment by DanielNoord** [comment]
I think you're misrepresenting some of the discussion here by claiming not a single valid example was given but also using your own interpretation of constant.

We have defined constants as being assigned to only once. We also understand that sometimes this is done in an `if ... else` or `try ... except` block and therefore consider those assignments constants as well. All other assignments are just regular variables: during execution of the program their value changes and you need to follow the logic of the program to understand what the value will be at the end of the program. In your example this is also the case: I can't determine the value of this variable from reading a single line of code but need to read several.
Of course this definition has its flaws, but lines need to be drawn somewhere.

The original issue received a lot of feedback from people considering `pylint` not to behave correctly. With this change we have fixed that at the expense of sometimes needing a disable comment in less than trivial "constant" assignments. We also nudge users to think more carefully about module constants and variables (by treating them differently) which I think is also a net positive and falls within the "responsibility" of a linter.

It is unfortunate you don't agree with the behaviour change but from a maintainers perspective this discussion has been quite taxing. It feels like you're not listening to what I and others are saying and can only see your own side of the story.

To help this discussion and hopefully make you understand my perspective better, let me explicitly share mine: 
`pylint` has always required the use of disable comments as some valid code just can't doesn't fit with generalised linting rules in combination with a somewhat sane amount of configuration (although you could argue the current number is far past that point). The code you shared seems like an example of this: the change in question seems valid for most cases but for more complex code you need a disable. I'm confirmed in this believe by other issues that were opened in which users mostly pointed out flaws in our "is reassigned" logic, not what the change itself.

That's my perspective. It might not be 100% correct but it does make me -0,5 like Pierre to revert this change or change more about it.

**Comment by Pierre-Sassoulas** [comment]
You have a strong opinion about this, we're not going to convince you that the new behavior is valid. I don't want to argue for hours about this which is why I linked to a claude AI answer to a neutral question asking it to compare the two pieces of code. 

We had a lot of comments from a lot of people complaining about pylint not being able to distinguish constant and module level variables properly and I just don't see this volume of complaints in this issue.

**Comment by avylove** [comment]
@DanielNoord I appreciate the feedback and I appreciate where you are coming from. The reason I keep pushing is the logic doesn't hold up to scrutiny, and when I bring up these points, they are not addressed.


> I think you're misrepresenting some of the discussion here by claiming not a single valid example was given but also using your own interpretation of constant.

If you look at all the examples in #3585, all but one are some variation of ``for i in ...:`` If you use this in a module and then use ``i`` in any function or methods in the module, pylint would raise a ``redefined-outer-name``. So you might argue this is just an example used to prove a point, not intended to be correct. OK, then can we look at a real world example?

The only other example is this:

```python
sym_perc_held = 1.0 # Problem line that gets flagged
if sym_exp_val > 0.0:
  sym_perc_held = sym_curr_val / sym_exp_val
```

This looks like either a constant or code that belongs in a function. It's hard to say without context.

> We have defined constants as being assigned to only once. ... Of course this definition has its flaws, but lines need to be drawn somewhere.

I get the line needs to be drawn somewhere, but it's arbitrary. If there was some elegant way to combine if and try statements, then we wouldn't be having this conversation. But we're working around the limitations of the language, including the fact that there is no real constant. (In other languages this would be resolved at compile time to a single assignment.) So I think we have to work backwards and say, "What does ideal code look like?", then try to get as close as we can to being able to identify it.


> We had a lot of comments from a lot of people complaining about pylint not being able to distinguish constant and module level variables properly and I just don't see this volume of complaints in this issue.

@Pierre-Sassoulas I see 6 people that aren't Pylint maintainers who commented on #2585 over the 5 years it was open. This change was made 2 weeks ago and this issue was open 3 days before being closed.

Pylint also gets a lot of issues where someone thinks Pylint has a false positive until they learn the logic behind it and realize they can improve their code. I suspect that is the primary driver behind #2585, but, again, it's hard to know without real examples.

I've said this before, but if we look at variables defined at the module root, they should fall into three buckets:
- Constants: Once the value is determine, it never changes
- Globals: May change after the module is initialized via the global keyword within a function. Rarely used. PEP8 says these should start with an underscore or ``__all__`` should be defined to prevent leakage outside of module
- Dunder main variables: Guarded by a dunder name clause

Everything else should be contained within a function or class.

My guess, without seeing real world examples, is people are leaving runtime logic in the root of their modules, perhaps in single file scripts. I've taught Python classes and Pylint usage to hundreds of people and that is extremely common. After I explain Python scripts are modules and talk a little about reuse, testing, etc, they usually get it and restructure their code.

> It is unfortunate you don't agree with the behaviour change but from a maintainers perspective this discussion has been quite taxing. It feels like you're not listening to what I and others are saying and can only see your own side of the story.

It is very taxing for me too. Each comment I leave takes about an hour. Every time I make a point, I have to think of a counter point and argue against myself, then revise. I am not "feeling", I am trying to get to a logical truth. I don't care whose truth it is as long as it's logically sound. So it's very frustrating when the logical arguments I make are completely ignored and we're just counting thumbs up.

**Comment by DanielNoord** [comment]
> If you look at all the examples in #3585

This is not about the examples given in the issue. The code that is affected is any module level variable assignment and whether they should be considered constants or variables. There are hundreds of examples to find in which this distinction is relevant. Especially for correctly structuring code, like you touch on below yourself.

> This looks like either a constant or code that belongs in a function. It's hard to say without context.

Agreed, unfortunately `pylint` needs to make a choice here as we can not suddenly ignore certain lines because we can't make a decision. In this case we would consider it to be a variable.

> I get the line needs to be drawn somewhere, but it's arbitrary. 

No disagreement here.

> So I think we have to work backwards and say, "What does ideal code look like?", then try to get as close as we can to being able to identify it.

This is just as arbitrary and context dependent as the points raised above.
At least with this change we're nudging users to think about the difference between a variable and constant and their effect on code readability. As I argued before I think that is a positive change.

> I've said this before, but if we look at variables defined at the module root, they should fall into three buckets:

Why? Why should they? This is just your definition of what they should fall into, or your interpretation of PEP 8. Please read Pierre's latest topic on the Python Discourse: PEP 8 is not authoritative and always up to interpretation given the circumstances.

Having said that, I like the distinctions you're making.

> - Constants: Once the value is determine, it never changes

How would `pylint` determine this? The current logic tries to do exactly this.

> - Globals: May change after the module is initialized via the global keyword within a function. Rarely used. PEP8 says these should start with an underscore or ``__all__`` should be defined to prevent leakage outside of module

There are many use cases for globals to be exposed. For example, the request context in a `flask` web app. If PEP 8 says this it is clear that those writing code have found use cases not covered by it.

> - Dunder main variables: Guarded by a dunder name clause
> 
> Everything else should be contained within a function or class.

This is simply not always true. People might want to define certain functions or classes based on certain logic. They might want to use variables to capture that logic. Just because you are not using variables at the module level doesn't mean others aren't. `pylint` needs to handle this.

With the last change at last `pylint` attempts to distinguish between the constants you have in your definition and the variables that occur in code.

> My guess, without seeing real world examples, is people are leaving runtime logic in the root of their modules, perhaps in single file scripts. I've taught Python classes and Pylint usage to hundreds of people and that is extremely common. After I explain Python scripts are modules and talk a little about reuse, testing, etc, they usually get it and restructure their code.

Agreed! With this change and `pylint`'s ability to distinguish between the two we are now also making users aware of this distinction. Whether they want to refactor it is up to them. We could make a check to nudge people to do so, but that's a different issue.

> I am not "feeling", I am trying to get to a logical truth. I don't care whose truth it is as long as it's logically sound.

I think the problem is that I and others think we will never find such a truth that is applicable to all codebases. We find one that is a good middle ground and which might require a disable here and there.

**Comment by avylove** [comment]
> This is not about the examples given in the issue. The code that is affected is any module level variable assignment and whether they should be considered constants or variables. There are hundreds of examples to find in which this distinction is relevant. Especially for correctly structuring code, like you touch on below yourself.

It just seems odd that no real world examples were given. If there are so many, finding a single example should be easy.

> > * Globals: May change after the module is initialized via the global keyword within a function. Rarely used. PEP8 says these should start with an underscore or `__all__` should be defined to prevent leakage outside of module
> 
> There are many use cases for globals to be exposed. For example, the request context in a `flask` web app. If PEP 8 says this it is clear that those writing code have found use cases not covered by it.

I'm not sure ``flask.request`` would match this definition of a global. It's really a constant both by my definition and the current Pylint implementation. It's assigned to a ``LocalProxy`` instance once in ``flask.globals`` and never changes.

> This is simply not always true. People might want to define certain functions or classes based on certain logic. They might want to use variables to capture that logic. Just because you are not using variables at the module level doesn't mean others aren't. `pylint` needs to handle this.

In theory, that is possible, but these variables sound like constants to me and Pylint will consider them constants if they are assigned once, but as variables if they assigned more than once.  The current logic distinguishing them based on how they are defined, not by how they are used. So I think the difference seems to be where the line is between use and definition. Understandable because both can and do happen at module initialization. But the values live on for the entire life of the program. My thinking is:

- Module initialization happens, the values are assigned, usually once, but maybe more depending on the logic
- If they change after module initialization or within the dunder name block, they are not constants

The current Pylint logic is
- If multiple assignments happen non-exclusively, anywhere in the module, it is not a constant

While I believe my thinking is more correct, I wonder if there is a way to improve the current logic to account for cases like mine. How hard would it be to see if the assignments were blocked together? Essentially supporting the case: assign the value, test the value, go to a fallback. But not support the case: assign the value, use the value, change the value. It's a subtle distinction. It would require assignment within a ``try`` block immediately after the initial assignment. Perhaps an ``if`` block would be valid too, but all the examples I could think of for ``if`` blocks were collapsible into the proceeding try/if/assignment.

**Comment by DanielNoord** [comment]
> It just seems odd that no real world examples were given. If there are so many, finding a single example should be easy.

I think this is because most of us are replying to you from our phone, so finding links to examples is a bit harder. But your own code I think is an example? It is a module level assignment that is more complex than a single line of code.

Other examples can be found in the PR that introduced this change.

> How hard would it be to see if the assignments were blocked together? 

That is what we're trying to do with support for try and if. Detecting when those are combined like in your code will likely be hard, but you could try. Recent PRs by @jacobtylerwalls show where the logic lives.

**Comment by avylove** [comment]
> > It just seems odd that no real world examples were given. If there are so many, finding a single example should be easy.
> 
> I think this is because most of us are replying to you from our phone, so finding links to examples is a bit harder. But your own code I think is an example? It is a module level assignment that is more complex than a single line of code.
> 
> Other examples can be found in the PR that introduced this change.
> 
> >

No, my example is a constant. I mean an example that has module level variables that aren't constants or guarded and are used in a way that's keeping with other good coding practices. Sure, it might be hard to find one on your phone, but the original issue was opened over 5 years ago and no real examples were ever presented. I feel like we are optimizing for something that doesn't exist.



> > How hard would it be to see if the assignments were blocked together?
> 
> That is what we're trying to do with support for try and if. Detecting when those are combined like in your code will likely be hard, but you could try. Recent PRs by [@jacobtylerwalls](https://github.com/jacobtylerwalls) show where the logic lives.

I'll take a look when I get a chance. It's been a while since I did anything with Astroid.

**Comment by Pierre-Sassoulas** [comment]
Believe me we had more than 5 persons complaining about invalid-name's behavior on module level variable here on github and elsewhere. There's stackoverflow posts about this that could be in high school by now. Some people want to name a logger, logger and not LOGGER and some the other way around. This is reasonable.

Wanting to reassign variable and still call them constants... This is less reasonable.

Basically, we disagree on the definition of what a constant is. You dismiss our (commonly admitted) definition and asks for an example of something that doesn't exist according to your own definition. I don't think this is a productive argument.

Also you claim the suggested fix of using a function is inferior like it's rational scientific facts when it's in fact akin to taste in poetry. (And ClaudeAI and I do not share your taste in poetry). 

This conversation is getting nowhere, please take at least a week to read about what other people think about this in legacy threads that led to this decision before answering again. Or create your own plugin as suggested before ?

**Comment by avylove** [comment]
> Believe me we had more than 5 persons complaining about invalid-name's behavior on module level variable here on github and elsewhere. There's stackoverflow posts about this that could be in high school by now. Some people want to name a logger, logger and not LOGGER and some the other way around. This is reasonable.

I've been looking for 3.5 days now for a real world example that would support the new behavior, but I haven't found a single example. Every use case seems to disappear when other good coding practices are applied. But I will do some crowd-sourcing and see if the community has anything.

I think I'm missing something. How does the new behavior change how logger is named?

> Also you claim the suggested fix of using a function is inferior like it's rational scientific facts when it's in fact akin to taste in poetry. (And ClaudeAI and I do not share your taste in poetry).

I gave a fair list of [pros and cons](https://github.com/pylint-dev/pylint/issues/10700#issuecomment-3456444205).

**Comment by jacobtylerwalls** [comment]
> I think I'm missing something. How does the new behavior change how logger is named?

I think @Pierre-Sassoulas is thinking of a slightly different part of the changes that were made in the same pull request.

The pull request did three main things:
- started checking type-annotated module names (these were totally skipped)
- started checking lists, sets, and objects (e.g. a logger) assigned to module names (these were totally skipped)
- started checking reassigned module names against the variable regex (this is the behavior being questioned now)

**Comment by avylove** [comment]
I posted [this](https://www.reddit.com/r/Python/comments/1oj4mcr/comment/nm6qgio/?sort=top) in r/Python a week ago. It got 59K views and 41 up votes. No one was able to provide any examples to support the new behavior. The most upvoted comment was:

> I'd see it as a safeguard against laziness or typos (= vs. ==).
> 
> And my guess is that the biggest beneficiary may be some proprietary, old codebase with heaps of technical debt where folks resort to hacks to get stuff done. :D

I also was unable to find any valid examples supporting the new behavior elsewhere.

I think this comment points to a general consensus that variables in the root of the a module should not change once determined and there should probably be a check for this. Globals do change, but this is done within a function/method via the global keyword. I think the limitations of the language still warrant an assign/try/fallback pattern, but I think that can be ignored manually if the rest of the logic was sound.

So here is my thinking as a more comprehensive strategy for module-level variables. 

- Error if any value in the module root is assigned more than once
  - While I would love if the assign/try/fallback pattern in the initial example was supported, I understand the complexities of this
- Any variable reassigned via the global keyword uses snake_case by default
- All other module-level variables not guarded by a dunder name clause are constants
  - Constants that are literals use UPPERCASE by default
  - A regex or boolean is added to allow users to determine how they want non-literal constants to be treated, defaulting to snake_case or UPPERCASE

**Comment by DanielNoord** [comment]
> I posted [this](https://www.reddit.com/r/Python/comments/1oj4mcr/comment/nm6qgio/?sort=top) in r/Python a week ago. 
> ...
> The most upvoted comment was:
> ...

You forgot to mention the comment that said moving your logic into a function as suggested by Pierre before would be better 😄 

> I think this comment points to a general consensus that variables in the root of the a module should not change once determined and there should probably be a check for this.

We already agreed on this before. Feel free to open an issue and suggest exact semantics, test cases, documentation examples, etc. Implementing new checks is generally quite a lot of work.

> So here is my thinking as a more comprehensive strategy for module-level variables.
> 
> * Error if any value in the module root is assigned more than once

Please let's keep this out of the discussion here. This is a separate check that should be discussed on its own merit.

>   * While I would love if the assign/try/fallback pattern in the initial example was supported, I understand the complexities of this
> * Any variable reassigned via the global keyword uses snake_case by default

Seems somewhat sensible to me. Would be interesting to see what other thinks.

> * All other module-level variables not guarded by a dunder name clause are constants

Still not so sure about this. With the current logic we're making it harder to do a lot of assignments to something that should be a constant in the end. Both Pierre and a random redditor have already suggested that they would actually prefer to have this logic in a separate function. The original issue also got some traction whereas we haven't closed a lot of duplicate issues for this one, so I guess not too many share your position?

At this point I'm really +-0 on this and tend to lean towards not another round of breaking changes.

>   * A regex or boolean is added to allow users to determine how they want non-literal constants to be treated, defaulting to snake_case or UPPERCASE


I'd like to avoid even more configuration options for naming things as it is already very complex for new users. But if this it the besy way forward so be it.

**Comment by avylove** [comment]
> You forgot to mention the comment that said moving your logic into a function as suggested by Pierre before would be better 😄

I also didn't mention the debate in the comments over of Pylint was still relevant because both are irrelevant to the purpose of the post, which was to solicit examples that support the new behavior. Also, if you recall, I was the one who initially proposed a function as a workaround.

> > I think this comment points to a general consensus that variables in the root of the a module should not change once determined and there should probably be a check for this.
> 
> We already agreed on this before. Feel free to open an issue and suggest exact semantics, test cases, documentation examples, etc. Implementing new checks is generally quite a lot of work.
> 
> > So here is my thinking as a more comprehensive strategy for module-level variables.
> > 
> > * Error if any value in the module root is assigned more than once
> 
> Please let's keep this out of the discussion here. This is a separate check that should be discussed on its own merit.
> 
> >

These are the same thing. The heart of the matter is the current implementation assumes if something is assigned more than once, it is a global or what you are calling a "module-level variable" and should be snake case. There is no pattern to support this logic. I have been unable to find one in two weeks, the original issue was open 5 years and had none, none of the Pylint maintainers have provided one, of the 60K viewers of the Reddit post, no one supplied one.

However, there is consensus that nothing in the root of the module should be reassigned. We have a little disagreement on if 
the assign/try/fallback is valid or not, but I am ok tabling that for the sake of simplicity.

> Both Pierre and a random redditor have already suggested that they would actually prefer to have this logic in a separate function

The same can be said of every example in #3585 , correct?

So I think the crux of it falls to two points:

- globals or "module-level variables" should not be determined based solely on reassignment, but rather if reassignment happens via the global keyword
- Multiple assignments at the module root, except when all assignments are within the dunder main block, should raise a warning

> Please let's keep this out of the discussion here. This is a separate check that should be discussed on its own merit.

I was thinking I'd crowd source refinements to what I proposed and then create an issue focused on a cohesive constants strategy.

**Comment by DanielNoord** [comment]
> * globals or "module-level variables" should not be determined based solely on reassignment, but rather if reassignment happens via the global keyword
> * Multiple assignments at the module root, except when all assignments are within the dunder main block, should raise a warning

```python
for i in (0, 1, 2):
    variable = i
```

Just this example alone to me shows that not everything in a module is necessary a constant. I don't think it is good practice to do these loops in the module scope but if you have disabled warnings about this (because you want to do logic in the module scope) it would be very confusing to get linter warnings about this not following to constant pattern.

The lack of other issues about this shows me that either:
- Users don't use `invalid-name`
- Users don't have complicated assignment patterns to module constants
- Users don't care about adding a `pylint: disable` on these lines

If you can find a solution that is satisfactory to you but also doesn't confuse new Python users with a confusing "Constant name ..." error for assignments that are clearly variables please propose a PR, but I haven't found one.

> > Please let's keep this out of the discussion here. This is a separate check that should be discussed on its own merit.
> 
> I was thinking I'd crowd source refinements to what I proposed and then create an issue focused on a cohesive constants strategy.

Thanks for this, it is always good to base a check on community input. That said, warnings should be evaluated on their own as they can all individually be ignored. I still think you have brought many good arguments for a warning that checks for "variable" assignments in the module level scope. Perhaps after we have written that we can leverage it for `invalid-name` again, but for now I think I'd like to focus my attention on issues that users care more about.

If you care enough about the issue discussed here yourself please open a PR. You have made a nice contribution to both `astroid` and `pylint` so I'd be more than happy to review an attempt that fixes this in a manner that is satisfactory for all parties.

**Comment by jacobtylerwalls** [comment]
I've been trying to absorb this discussion slowly.

I'm still open to adding a new global name regex as I mentioned here: https://github.com/pylint-dev/pylint/issues/10700#issuecomment-3448648507

@avylove seemed in favor directly after (https://github.com/pylint-dev/pylint/issues/10700#issuecomment-3448703152). @DanielNoord was open to it at the end of https://github.com/pylint-dev/pylint/issues/10700#issuecomment-3493196084. @Pierre-Sassoulas was skeptical at first (https://github.com/pylint-dev/pylint/issues/10700#issuecomment-3455676045):

> While I believe more regex options might be the way to go for pylint... Adding a regex options will introduce negative side effects that we also need to consider. What happens if someone already use the current regex and we introduce the new one ?

We should do this relatively quickly so that we don't have too much churn as people slowly upgrade to 4. It's only been a few weeks. We need to admit if a new feature is causing false positives. We didn't do a pre-release because we didn't get meaningful feedback last time. @avylove did the right thing by raising it as quickly as possible. 

> The lack of other issues about this shows me that either:

> Users don't use invalid-name

That might be it. Among the projects we test in the pylint primer, these projects disable it:

- astroid
- pytest
- coverage
- music21
- ...

I stopped looking after that.

As I mentioned in https://github.com/pylint-dev/pylint/issues/10700#issuecomment-3448648507, we should have made the new behavior opt-in. We can still do that with a regex.

Can I ask @Pierre-Sassoulas to give this some thought? 🙏

**Comment by Pierre-Sassoulas** [comment]
I just found a useless redefinition of a constant by upgrading pylint thanks to the new behavior, so I like it.

There's value in having clear definition of thing and sticking to it. A constant is ... Constant and shouldn't be redefined, it's a pretty basic and reasonable definition. If you follow the definition you're going to benefit when assumptions are broken. If you're not following the definition you can't benefit from automated lint. 

If we have to change something let's not make invalid-name even more complex with a hard to handle breaking change for regexes users, and just allow both style by default for module level variable with an option like '--module-level-variable-invalid-name-check'. We can't detect automatically if something is intended to be a constant so we're never going to not have false negatives / false positives for this as long as we try to distinguish between the two. (As you know, I've been a long time advocate of this. The argument against it was it's disruptive afair, and I think it's less disruptive than adding another regex). Or do both, because if the new regex is opt in, only those who want more control will have to deal with it after asking for module level variable check.

I also think the check in its current state is reasonable and that there's (a lot of) more pressing matter in pylint / astroid.

**Comment by avylove** [comment]
> The lack of other issues about this shows me that either:
> 
> * Users don't use `invalid-name`
> * Users don't have complicated assignment patterns to module constants
> * Users don't care about adding a `pylint: disable` on these lines

It's all of the above. I've talked to a lot of people about this in the last few weeks and a lot of either don't have much in the module root or use in-line disable when an error is raised.

> There's value in having clear definition of thing and sticking to it. A constant is ... Constant and shouldn't be redefined, it's a pretty basic and reasonable definition. If you follow the definition you're going to benefit when assumptions are broken. If you're not following the definition you can't benefit from automated lint.

I think we've moved past this. I'm now ok with the concept that ideally constants shouldn't be reassigned. For legitimate cases, like my initial example, they can be moved to a function or in-line disabled. But the same should apply to globals. I think using reassignment as a differentiator is  sometimes going to give you a good signal and sometimes won't depending on how you named things. I agree flaging reassignment is useful, but it's less useful if the check is dependent on naming conventions. There's also no difference for variables guarded by dunder name.

This currently doesn't raise a warning.

```
if __name__ == '__main__':
    VARIABLE = 'something'
```

But this does
```
if __name__ == '__main__':
    VARIABLE = 'something'
    VARIABLE = 'something_else'
```

And this does, but it doesn't warn you that you reassigned a constant, but that it should be in snake case

```
VARIABLE = 'something'
if __name__ == '__main__':
    VARIABLE = 'something_else'
```

Is there any legitimate reason to define a constant under dunder name? It seems like a code smell.

What I'd like to see is

- If ALL assignments for a variable are under dunder name , it should be treated as a normal variable (variable-rgx)
- A warning for ANYTHING defined outside of dunder name that is reassigned
- Anything reassigned via the `global` keyword treated as a global. To match PEP8, this should probably have it's own regex
- A new regex for non-literal constants since they are no longer ignored

Given use of global is not common and non encouraged, the third one is probably lower priority.
